src/Mesh/DerivedData.C

```diff
@@ -801,4 +802,311 @@ genEsued( const std::vector< std::size_t >& inpoel,
+              if (lpoin[jpoin] == 1) { ++icoun; }
+            }
+            //store esuel if
+            if (icoun == nnpf)
+            {
+              auto markf = nfpe*e;
+              esuelTet[markf+fe] = static_cast<int>(jelem);
+
+              markf = nfpe*jelem;
+              esuelTet[markf+fj] = static_cast<int>(e);
+            }
+          }
+        }
+      }
+      // reset this array
+      lpoin[lhelp[0]] = 0;
+      lpoin[lhelp[1]] = 0;
+      lpoin[lhelp[2]] = 0;
+    }
+  }
+
+  return esuelTet;
+}
+
+std::size_t 
+genNtfac( std::size_t nfpe,
+          const std::size_t& nbfac,
+          const std::vector< int >& esuelTet )
+// *****************************************************************************
+//  Generate derived data, total number of faces in the mesh
+//  added by Aditya K Pandare
+//! \param[in] nfpe Number of faces per element.
+//! \param[in] nbfac Number of boundary faces.
+//! \param[in] esuelTet Elements surrounding elements.
+//! \return Total number of faces in the mesh
+//! \details The unsigned integer here gives the total number of faces in 
+//     the mesh.
+// *****************************************************************************
+{
+  Assert( !esuelTet.empty(), "Attempt to call genNtfac() with empty esuelTet" );
+  Assert( esuelTet.size()%nfpe == 0, 
+                  "Size of esuelTet must be divisible by nfpe" );
+  Assert( nfpe > 0, "Attempt to call genNtfac() with zero faces per element" );
+
+  auto nelem = esuelTet.size()/nfpe;
+
+  std::size_t ntfac(0), nifac(0);
+  // note: nbfac is assumed to be already initialized
+
+  // loop through elements surrounding elements to find number of internal faces
+  for (std::size_t e=0; e<nelem; ++e)
+  {
+    for (std::size_t ip=nfpe*e; ip<nfpe*(e+1); ++ip)
+    {
+      if (esuelTet[ip] != -1)
+      {
+        if ( e<static_cast< std::size_t >(esuelTet[ip]) )
+        {
+          ++nifac;
+        }
+      }
+    }
+  }
+
+  ntfac = nifac + nbfac;
+
+  return ntfac;
+}
+
+std::vector< int >
+genEsuf( std::size_t nfpe,
+         const std::size_t& ntfac,
+         const std::size_t& nbfac,
+         const std::map< int, std::vector< std::size_t > >& belem,
+         const std::vector< int >& esuelTet )
+// *****************************************************************************
+//  Generate derived data, elements surrounding faces
+//  added by Aditya K Pandare
+//! \param[in] nfpe  Number of faces per element.
+//! \param[in] ntfac Total number of faces.
+//! \param[in] nbfac Number of boundary faces.
+//! \param[in] belem Boundary element map according to side-sets.
+//! \param[in] esuelTet Elements surrounding elements.
+//! \return Elements surrounding faces.
+//! \details The unsigned integer vector gives the IDs of the elements to the 
+//    left and the right of each face in the mesh. The convention followed 
+//    throughout is : The left element always has an ID smaller than the ID of 
+//    the right element.
```

Comment:
Thanks, that answers my question on how you guys define the "left" and "right" elements. Makes sense.

Reply:
Cool!